sklearn/preprocessing/data.py

```diff
@@ -1113,39 +1101,61 @@ def fit(self, X, y=None):
             The data used to compute the median and quantiles
             used for later scaling along the features axis.
         """
-        if sparse.issparse(X):
-            raise TypeError("RobustScaler cannot be fitted on sparse inputs")
-        X = self._check_array(X, self.copy)
+        # at fit, convert sparse matrices to csc for optimized computation of
+        # the quantiles
+        X = check_array(X, accept_sparse='csc', copy=self.copy, estimator=self,
+                        dtype=FLOAT_DTYPES, force_all_finite='allow-nan')
+
+        q_min, q_max = self.quantile_range
+        if not 0 <= q_min <= q_max <= 100:
+            raise ValueError("Invalid quantile range: %s" %
+                             str(self.quantile_range))
+
         if self.with_centering:
-            self.center_ = np.median(X, axis=0)
+            if sparse.issparse(X):
+                raise ValueError(
+                    "Cannot center sparse matrices: use `with_centering=False`"
+                    " instead. See docstring for motivation and alternatives.")
+            self.center_ = nanmedian(X, axis=0)
+        else:
+            self.center_ = None
 
         if self.with_scaling:
-            q_min, q_max = self.quantile_range
-            if not 0 <= q_min <= q_max <= 100:
-                raise ValueError("Invalid quantile range: %s" %
-                                 str(self.quantile_range))
+            quantiles = []
+            for feature_idx in range(X.shape[1]):
+                if sparse.issparse(X):
+                    column_nnz_data = X.data[X.indptr[feature_idx]:
+                                             X.indptr[feature_idx + 1]]
+                    column_data = np.zeros(shape=X.shape[0], dtype=X.dtype)
+                    column_data[:len(column_nnz_data)] = column_nnz_data
+                else:
+                    column_data = X[:, feature_idx]
 
-            q = np.percentile(X, self.quantile_range, axis=0)
-            self.scale_ = (q[1] - q[0])
+                quantiles.append(
+                    nanpercentile(column_data, self.quantile_range)
+                    if column_data.size else [0, 0])
```

Comment:
I think you no longer need this `if column_data.size` check?

Reply:
nop because it returns nan otherwise (which is consistent with the numpy function).

```python
In [1]: import numpy as np

In [2]: from sklearn.utils.fixes import nanpercentile

In [3]: nanpercentile([], [25, 50, 75])
/home/lemaitre/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/lib/nanfunctions.py:1148: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
Out[3]: nan

In [4]: np.nanpercentile([], [25, 50, 75])
/home/lemaitre/miniconda3/envs/dev/lib/python3.6/site-packages/numpy/lib/nanfunctions.py:1148: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
Out[4]: nan
```

Reply:
But when would you encounter 0 samples after check_array?

Reply:
With sparse matrices with all zero column:

```python
In [1]: from scipy import sparse

In [3]: from sklearn.utils import check_array

In [4]: X = sparse.rand(10, 2, density=0).tocsr()

In [5]: check_array(X, accept_sparse=True)
Out[5]: 
<10x2 sparse matrix of type '<class 'numpy.float64'>'
	with 0 stored elements in Compressed Sparse Row format>

In [6]: X.data
Out[6]: array([], dtype=float64)

```

Reply:
Oh sorry, I see now

```
column_data = np.zeros(shape=X.shape[0], dtype=X.dtype)
```

I missed this and I was thinking that we were using the `data.nnz` which can be an empty array.

Making the changes then.